File: Backend/Controller/notes.controller.js
// controllers/noteController.js
import Notes from '../Model/notes.model.js'; // Adjust the path if necessary
import User from '../Model/user.model.js';

// Create a new note
export const createNote = async (req, res) => {
  try {
    const { title, content, tags, userId } = req.body;
    console.log(` Title: ${title}, Content: ${content}, Tags: ${tags}, userId: ${userId}`);
    
    const existingUser = await User.findById(userId); 
    if (!existingUser) {
      res.status(400).json({ message: 'User not Found'});
    }

    const newNote = new Notes({ title, content, tags, userId });
    await newNote.save();
    res.status(201).json({ message: 'Note created successfully', note: newNote });
  } catch (error) {
    console.log('Error creating note: ',error);
    
    res.status(500).json({ message: 'Error creating note', error: error.message });
  }
};

// Get all notes
export const getAllNotes = async (req, res) => {
  try {
    const { userId } = req.params; // Destructure userId from req.params

    const notes = await Notes.find({ userId }); // Find notes that match the userId

    res.status(200).json(notes);
  } catch (error) {
    res.status(500).json({ message: 'Error fetching notes', error: error.message });
  }
};


// Update a note
export const updateNote = async (req, res) => {
  try {
    const { title, content, tags } = req.body;
    const note = await Notes.findByIdAndUpdate(
      req.params.id,
      { title, content, tags },
      { new: true, runValidators: true }
    );

    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }

    res.status(200).json({ message: 'Note updated successfully', note });
  } catch (error) {
    res.status(500).json({ message: 'Error updating note', error: error.message });
  }
};

// Delete a note
export const deleteNote = async (req, res) => {
  try {
    const note = await Notes.findByIdAndDelete(req.params.id);
    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }
    res.status(200).json({ message: 'Note deleted successfully' });
  } catch (error) {
    res.status(500).json({ message: 'Error deleting note', error: error.message });
  }
};

export const getNote = async (req, res) => {
  try {
    const note = await Notes.findById(req.params.id)
    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }
    res.status(200).json(note);
  } catch (error) {
    res.status(500).json({ message: 'Error fetching a note', error: error.message });
  }
}

File: lib/Controller/notes.controller.dart
import 'package:get/get.dart';
import 'package:mindmap/Model/note.model.dart';
import 'package:mindmap/Service/Remote.services.dart';

class NoteController extends GetxController {
  var notes = <Note>[].obs;  // Changed variable name to lowercase for consistency
  var isLoading = true.obs;
  Rx<Note?> selectedNote = Rx<Note?>(null);

  @override
  void onInit() {
    fetchNotes();
    super.onInit();
  }

  Future<void> fetchNotes() async {
    isLoading(true);
    try {
      var fetchedNotes = await RemoteServices.fetchNotes();
      notes.value = fetchedNotes;
      print("Note Array: ${notes}");
    } finally {
      isLoading(false);
    }
  }

Future<void> fetchNoteById(String id) async {
  isLoading(true);
  try {
    Note note = await RemoteServices.fetchNoteById(id);
    selectedNote.value = note;
  } catch (e) {
    print("Error fetching note: $e");
  } finally {
    isLoading(false);
  }
}

  // Method to edit a note
  Future<void> editNote(String noteId, String title, String content, List<String> tags) async {
    isLoading(true);
    try {
      // Call the service method to update the note
      var updatedNote = await RemoteServices.editNote(noteId, title, content, tags);
      
      // Find the index of the note being updated
      final index = notes.indexWhere((note) => note.id == noteId); // Assuming note.id is the identifier

      if (index != -1) {
        // Update the existing note in the list
        notes[index] = notes.last;
      }
    } catch (e) {
      // Handle errors here (e.g., show a message to the user)
      print('Error updating note: $e');
    } finally {
      isLoading(false);
    }
  }

    void deleteNote(String noteId) async {
    try {
      await RemoteServices.deleteNotes(noteId); // Call the delete method in RemoteServices
      notes.removeWhere((note) => note.id == noteId); // Remove the note from the local list
      Get.snackbar('Success', 'Note deleted successfully', // Show a success message
          snackPosition: SnackPosition.BOTTOM);
    } catch (e) {
      print("Error deleting note: $e"); // Handle error appropriately
      Get.snackbar('Error', 'Failed to delete note', // Show an error message
          snackPosition: SnackPosition.BOTTOM);
    }
  }

  Future<void> addNote(String title, String content, List<String> tags) async {
    try {
      isLoading(true);
      Note newNote = Note(
        id: '', // Assuming the server generates this ID
        title: title,
        content: content,
        tags: tags,
      );

      // Call remote service to add note
      Note addedNote = await RemoteServices.createNote(title, content, tags);
      notes.add(addedNote); // Update local notes list
    } catch (e) {
      throw Exception('Error adding note: ${e.toString()}');
    } finally {
      isLoading(false);
    }
  }

}
